File: golang-http-server/internal/service/attachment_service.go
package service

import (
	"context"
	"fmt"

	"golang-http-server/internal/core"
	"golang-http-server/internal/core/errors"
	"golang-http-server/internal/db/models"
	"golang-http-server/internal/dto"
	"golang-http-server/internal/repository"
)

type AttachmentService struct {
	repo     *repository.AttachmentRepository
	s3Client *core.S3Client
}

func NewAttachmentService(repo *repository.AttachmentRepository, s3Client *core.S3Client) *AttachmentService {
	return &AttachmentService{
		repo:     repo,
		s3Client: s3Client,
	}
}

func (s *AttachmentService) List(ctx context.Context, taskID string, limit, offset int) ([]dto.AttachmentResponse, core.PaginationMeta, error) {
	logger := core.GetLogger()
	logger.Debug("listing attachments", "task_id", taskID, "limit", limit, "offset", offset)

	attachments, total, err := s.repo.ListByTask(ctx, taskID, limit, offset)
	if err != nil {
		return nil, core.PaginationMeta{}, err
	}

	meta := core.BuildPaginationMeta(total, limit, offset)
	responses := make([]dto.AttachmentResponse, len(attachments))
	for i, a := range attachments {
		responses[i] = attachmentToResponse(a)
	}

	return responses, meta, nil
}

func (s *AttachmentService) FindByID(ctx context.Context, id string) (*dto.AttachmentResponse, error) {
	logger := core.GetLogger()
	logger.Debug("finding attachment by ID", "id", id)

	attachment, err := s.repo.GetByID(ctx, id)
	if err != nil {
		return nil, err
	}
	if attachment == nil {
		return nil, errors.NewNotFoundError("Attachment")
	}

	response := attachmentToResponse(*attachment)
	return &response, nil
}

func (s *AttachmentService) Create(ctx context.Context, taskID string, params dto.AttachmentCreate) (*dto.AttachmentUploadResponse, error) {
	logger := core.GetLogger()
	logger.Info("creating attachment", "file_name", params.FileName, "task_id", taskID)

	if err := params.Validate(); err != nil {
		return nil, err
	}

	fileKey := fmt.Sprintf("tasks/%s/%s", taskID, params.FileName)

	attachment, err := s.repo.Create(ctx, taskID, params.FileName, fileKey, params.MimeType, params.FileSize, nil)
	if err != nil {
		return nil, err
	}

	uploadURL, err := s.s3Client.GetPresignedPutURL(fileKey, 3600)
	if err != nil {
		return nil, fmt.Errorf("get presigned upload URL: %w", err)
	}

	response := attachmentToResponse(*attachment)
	return &dto.AttachmentUploadResponse{
		UploadURL:  uploadURL,
		Attachment: response,
	}, nil
}

func (s *AttachmentService) GetDownloadURL(ctx context.Context, id string) (string, error) {
	logger := core.GetLogger()
	logger.Info("getting download URL", "id", id)

	attachment, err := s.repo.GetByID(ctx, id)
	if err != nil {
		return "", err
	}
	if attachment == nil {
		return "", errors.NewNotFoundError("Attachment")
	}

	downloadURL, err := s.s3Client.GetPresignedURL(attachment.FileKey, 3600)
	if err != nil {
		return "", fmt.Errorf("get presigned download URL: %w", err)
	}

	return downloadURL, nil
}

func (s *AttachmentService) Delete(ctx context.Context, id string) error {
	logger := core.GetLogger()
	logger.Info("deleting attachment", "id", id)

	exists, err := s.repo.Exists(ctx, id)
	if err != nil {
		return err
	}
	if !exists {
		return errors.NewNotFoundError("Attachment")
	}

	return s.repo.Delete(ctx, id)
}

func (s *AttachmentService) HealthCheck(ctx context.Context) bool {
	count, err := s.repo.Count(ctx, "")
	return err == nil && count >= 0
}

func attachmentToResponse(a models.Attachment) dto.AttachmentResponse {
	resp := dto.AttachmentResponse{
		ID:        a.ID,
		TaskID:    a.TaskID,
		FileName:  a.FileName,
		FileKey:   a.FileKey,
		MimeType:  a.MimeType,
		FileSize:  a.FileSize,
		CreatedAt: a.CreatedAt,
	}
	if a.UploadedBy.Valid {
		resp.UploadedBy = &a.UploadedBy.String
	}
	return resp
}

File: fastapi/src/services/attachment_service.py
import uuid
from uuid import UUID

from fastapi import Depends
from sqlalchemy.ext.asyncio import AsyncSession

from src.core.config import Settings, get_settings
from src.core.errors.not_found_error import NotFoundError
from src.core.logger import get_logger
from src.core.pagination import PaginationMeta, build_pagination_meta
from src.core.s3_client import S3Client, get_s3_client
from src.db import get_db
from src.db.models.attachment import Attachment
from src.repositories.attachment_repository import AttachmentRepository, get_attachment_repository
from src.schemas.attachment_schemas import AttachmentCreate


class AttachmentService:
    def __init__(self, repository: AttachmentRepository, s3_client: S3Client) -> None:
        self._repository = repository
        self._s3_client = s3_client

    async def list(
        self,
        *,
        task_id: UUID,
        limit: int = 20,
        offset: int = 0,
    ) -> tuple[list[Attachment], PaginationMeta]:
        logger = get_logger()
        logger.debug("Listing attachments", task_id=str(task_id), limit=limit, offset=offset)

        items, total = await self._repository.list_by_task(task_id=str(task_id), limit=limit, offset=offset)

        meta = build_pagination_meta(total, limit, offset)
        return items, meta

    async def find_by_id(self, id: UUID) -> Attachment:
        logger = get_logger()
        logger.debug("Finding attachment by ID", id=str(id))

        attachment = await self._repository.get_by_id(id)
        if attachment is None:
            raise NotFoundError("Attachment")
        return attachment

    async def create(
        self,
        task_id: UUID,
        params: AttachmentCreate,
        uploaded_by: UUID | None = None,
    ) -> tuple[Attachment, str]:
        logger = get_logger()
        logger.info("Creating attachment", file_name=params.file_name)

        file_key = f"uploads/{uuid.uuid4()}-{params.file_name}"

        attachment = await self._repository.create(
            id=uuid.uuid4(),
            task_id=task_id,
            file_name=params.file_name,
            file_key=file_key,
            mime_type=params.mime_type,
            file_size=params.file_size,
            uploaded_by=uploaded_by,
        )

        upload_url = self._s3_client.get_presigned_put_url(file_key)

        return attachment, upload_url

    async def get_download_url(self, id: UUID, expires_in_sec: int = 3600) -> str:
        attachment = await self.find_by_id(id)
        return self._s3_client.get_presigned_url(attachment.file_key, expires_in_sec)

    async def delete(self, id: UUID) -> bool:
        logger = get_logger()
        logger.info("Deleting attachment", id=str(id))

        attachment = await self.find_by_id(id)
        await self._repository.delete(attachment)
        return True

    async def health_check(self) -> bool:
        try:
            await self._repository.count()
            return True
        except Exception:
            return False


def get_attachment_service(
    session: AsyncSession = Depends(get_db),
    settings: Settings = Depends(get_settings),
) -> AttachmentService:
    repository = get_attachment_repository(session)
    s3_client = get_s3_client(settings)
    return AttachmentService(repository, s3_client)
